common/map_coordinate.h

```diff
@@ -0,0 +1,231 @@
+}
+
+// Transform cartesian coordinates from WGS84 into desired space
+// Ported from Javascript Ordnance Survey Grid Reference functions (c) Chris Veness 2005-2017 (MIT Licence)
+template<typename Datum>
+Cartesian<Datum> transformCartesian(const Cartesian<WGS84> &c)
+{
+    const double s1 = Datum::transform.s / 1E6 + 1.0;            // scale: normalise parts-per-million to (s+1)
+
+    const double rx = Datum::transform.rx.value();
+    const double ry = Datum::transform.ry.value();
+    const double rz = Datum::transform.rz.value();
+
+    // apply transform
+    return Cartesian<Datum>{Datum::transform.tx + c.x * s1 - c.y * rz + c.z * ry,
+                            Datum::transform.ty + c.x * rz + c.y * s1 - c.z * rx,
+                            Datum::transform.tz - c.x * ry + c.y * rx + c.z * s1};
+}
+
+// Convert cartesian coordinates back to latitude and longitude
+// Ported from Javascript Ordnance Survey Grid Reference functions (c) Chris Veness 2005-2017 (MIT Licence)
+template<typename Datum>
+LatLon<Datum> cartesianToLatLon(const Cartesian<Datum> &c)
+{
+    using namespace units::angle;
+    using namespace units::length;
+    using namespace units::math;
+    using namespace units::literals;
+
+    const double e2 = 2.0 * Datum::ellipsoid.f - Datum::ellipsoid.f * Datum::ellipsoid.f;    // 1st eccentricity squared ≡ (a²-b²)/a²
+    const double epsilon2 = e2 / (1.0 - e2);                                                // 2nd eccentricity squared ≡ (a²-b²)/b²
+    const meter_t p = sqrt(c.x * c.x + c.y * c.y);  // distance from minor axis
+    const meter_t r = sqrt(p * p + c.z * c.z);      // polar radius
+
+    // parametric latitude (Bowring eqn 17, replacing tanβ = z·a / p·b)
+    const double tanBeta = (Datum::ellipsoid.b * c.z) / (Datum::ellipsoid.a * p) * (1.0 + epsilon2 * Datum::ellipsoid.b / r);
+    const double sinBeta = tanBeta / std::sqrt(1.0 + tanBeta * tanBeta);
+    const double cosBeta = sinBeta / tanBeta;
+
+    // geodetic latitude (Bowring eqn 18: tanφ = z+ε²bsin³β / p−e²cos³β)
+    const degree_t phi = std::isnan(cosBeta) ? 0.0_rad : atan2(c.z + epsilon2 * Datum::ellipsoid.b * sinBeta * sinBeta * sinBeta,
+                                                               p - e2 * Datum::ellipsoid.a * cosBeta * cosBeta * cosBeta);
+
+    // longitude
+    const degree_t lambda = atan2(c.y, c.x);
+
+    // height above ellipsoid (Bowring eqn 7) [not currently used]
+    /*const double sinPhi = sin(phi), cosPhi = cos(phi);
+    const meter_t nu = std::get<0>(ellipsoid) / std::sqrt(1.0 - e2 * sinPhi * sinPhi); // length of the normal terminated by the minor axis
+    var h = p*cosφ + cartesian[2]*sinφ - (osgb36EllipseA*osgb36EllipseA/ν);*/
+
+    return LatLon<Datum>{phi, lambda};
+}
+
+// Convert latitude and longitude in the OSGB36 space to OS grid coordinates
+// Ported from Javascript Ordnance Survey Grid Reference functions (c) Chris Veness 2005-2017 (MIT Licence)
```

Comment:
Would you mind doing these comments in doxygen style? I.e.:
```
/**!
 * \brief Convert latitude and longitude in the OSGB36 space to OS grid coordinates
 * Ported from Javascript Ordnance Survey Grid Reference functions (c) Chris Veness 2005-2017 (MIT Licence)
 */

Reply:
Yeah it does, but I think ``//! Comments`` are for single-liners: i.e. just a brief description. The example I gave is one line of brief description (i.e. in the doxygen index) + extra info on the main page.

Reply:
...I don't have a preference style-wise though.